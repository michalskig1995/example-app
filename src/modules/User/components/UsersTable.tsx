import { useDispatch } from "react-redux";
import { useNavigate } from "react-router-dom";

import { IconPencil, IconTrash } from "@tabler/icons-react";
import { ActionIcon, Anchor, Avatar, Group, Table, Text } from "@mantine/core";

import { deleteUser } from "../../../modules/User/redux";

import type { Dispatch } from "redux";
import type { UserActionTypes } from "../../../modules/User/redux";
import type { User } from "../../../modules/User/types/User";

type UserTableProps = {
  users: User[];
};

export function UsersTable(props: UserTableProps) {
  const { users } = props;
  const dispatch = useDispatch<Dispatch<UserActionTypes>>();
  const navigate = useNavigate();

  function removeUser(userId: number) {
    dispatch(deleteUser(userId));
  }

  const rows = users.map((user) => (
    <Table.Tr key={user.name}>
      <Table.Td>
        <Group gap="sm">
          <Avatar
            size={30}
            src={`https://raw.githubusercontent.com/mantinedev/mantine/master/.demo/avatars/avatar-${user.id}.png`}
            radius={30}
          />
          <Text fz="sm" fw={500}>
            {user.name}
          </Text>
        </Group>
      </Table.Td>
      <Table.Td>
        <Anchor component="button" size="sm">
          {user.email}
        </Anchor>
      </Table.Td>
      <Table.Td>
        <Text fz="sm">{user.phone}</Text>
      </Table.Td>
      <Table.Td>
        <Group gap={0} justify="flex-end">
          <ActionIcon
            variant="subtle"
            color="gray"
            onClick={() => {
              navigate(`/users/${user.id}`);
            }}
          >
            <IconPencil size={16} stroke={1.5} />
          </ActionIcon>
          <ActionIcon
            variant="subtle"
            color="red"
            onClick={() => {
              removeUser(user.id);
            }}
          >
            <IconTrash size={16} stroke={1.5} />
          </ActionIcon>
        </Group>
      </Table.Td>
    </Table.Tr>
  ));

  return (
    <Table.ScrollContainer minWidth={800}>
      <Table verticalSpacing="sm">
        <Table.Thead>
          <Table.Tr>
            <Table.Th>Employee</Table.Th>
            <Table.Th>Email</Table.Th>
            <Table.Th>Phone</Table.Th>
            <Table.Th />
          </Table.Tr>
        </Table.Thead>
        <Table.Tbody>{rows}</Table.Tbody>
      </Table>
    </Table.ScrollContainer>
  );
}
