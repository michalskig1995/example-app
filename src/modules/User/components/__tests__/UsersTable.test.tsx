import { Provider } from "react-redux";

import { configureStore } from "@reduxjs/toolkit";
import { render } from "@testing-library/react";
import { MantineProvider } from "@mantine/core";

import { mockUsers } from "../../data";
import { UsersTable } from "../UsersTable";

jest.mock("redux-observable", () => ({
  createEpicMiddleware: jest.fn().mockReturnValue({
    run: jest.fn(),
  }),
}));

jest.mock("react-router-dom", () => ({
  useNavigate: jest.fn(),
}));

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: jest.fn(),
}));

jest.mock("@mantine/core", () => {
  const actualMantine = jest.requireActual("@mantine/core");

  return {
    ...actualMantine,
    Table: {
      ...actualMantine.Table,
      ScrollContainer: jest.fn(({ children }) => <div>{children}</div>),
      Thead: jest.fn(({ children }) => <thead>{children}</thead>),
      Tr: jest.fn(({ children }) => <tr>{children}</tr>),
      Th: jest.fn(({ children }) => <th>{children}</th>),
      Tbody: jest.fn(({ children }) => <tbody>{children}</tbody>),
      Td: jest.fn(({ children }) => <td>{children}</td>),
    },
    Group: jest.fn(({ children }) => <div>{children}</div>),
    Avatar: jest.fn(() => <div data-testid="mock-avatar" />),
    ActionIcon: jest.fn(({ children, onClick }) => (
      <button onClick={onClick} data-testid="mock-action-icon">
        {children}
      </button>
    )),
    Text: jest.fn(({ children }) => <span>{children}</span>),
    Anchor: jest.fn(({ children }) => <a>{children}</a>),
  };
});

const initialState = {
  users: mockUsers,
};

const mockStore = configureStore({
  reducer: (state = initialState) => state,
});

function renderWrapper() {
  return render(
    <Provider store={mockStore}>
      <MantineProvider theme={{}}>
        <UsersTable users={mockUsers} />
      </MantineProvider>
    </Provider>
  );
}

test("renders the users table with user data", () => {
  const { container } = renderWrapper();

  expect(container).toMatchSnapshot();
});
