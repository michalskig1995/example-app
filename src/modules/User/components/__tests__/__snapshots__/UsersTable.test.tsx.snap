// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the users table with user data 1`] = `
<div>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div>
    <table
      class="m_b23fa0ef mantine-Table-table"
      style="--table-vertical-spacing: var(--mantine-spacing-sm);"
    >
      <thead>
        <tr>
          <th>
            Employee
          </th>
          <th>
            Email
          </th>
          <th>
            Phone
          </th>
          <th />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <div>
              <div
                data-testid="mock-avatar"
              />
              <span>
                Leanne Graham
              </span>
            </div>
          </td>
          <td>
            <a>
              Sincere@april.biz
            </a>
          </td>
          <td>
            <span>
              1-770-736-8031 x56442
            </span>
          </td>
          <td>
            <div>
              <button
                data-testid="mock-action-icon"
              >
                <svg
                  class="tabler-icon tabler-icon-pencil "
                  fill="none"
                  height="16"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4"
                  />
                  <path
                    d="M13.5 6.5l4 4"
                  />
                </svg>
              </button>
              <button
                data-testid="mock-action-icon"
              >
                <svg
                  class="tabler-icon tabler-icon-trash "
                  fill="none"
                  height="16"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4 7l16 0"
                  />
                  <path
                    d="M10 11l0 6"
                  />
                  <path
                    d="M14 11l0 6"
                  />
                  <path
                    d="M5 7l1 12a2 2 0 0 0 2 2h8a2 2 0 0 0 2 -2l1 -12"
                  />
                  <path
                    d="M9 7v-3a1 1 0 0 1 1 -1h4a1 1 0 0 1 1 1v3"
                  />
                </svg>
              </button>
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div>
              <div
                data-testid="mock-avatar"
              />
              <span>
                Ervin Howell
              </span>
            </div>
          </td>
          <td>
            <a>
              Shanna@melissa.tv
            </a>
          </td>
          <td>
            <span>
              010-692-6593 x09125
            </span>
          </td>
          <td>
            <div>
              <button
                data-testid="mock-action-icon"
              >
                <svg
                  class="tabler-icon tabler-icon-pencil "
                  fill="none"
                  height="16"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4"
                  />
                  <path
                    d="M13.5 6.5l4 4"
                  />
                </svg>
              </button>
              <button
                data-testid="mock-action-icon"
              >
                <svg
                  class="tabler-icon tabler-icon-trash "
                  fill="none"
                  height="16"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4 7l16 0"
                  />
                  <path
                    d="M10 11l0 6"
                  />
                  <path
                    d="M14 11l0 6"
                  />
                  <path
                    d="M5 7l1 12a2 2 0 0 0 2 2h8a2 2 0 0 0 2 -2l1 -12"
                  />
                  <path
                    d="M9 7v-3a1 1 0 0 1 1 -1h4a1 1 0 0 1 1 1v3"
                  />
                </svg>
              </button>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
